### ISSUE TITLE
FTBFS on Fedora 22: ‘JPEG_GREYSCALE’ was not declared

### ISSUE THREAD

**Issue by eyalroz** [arrayfire/arrayfire#1360]
After supposedly having obtained all of the required libraries, I started the build. At some point I get this:

```
[ 25%] Building CXX object src/backend/cuda/CMakeFiles/afcuda.dir/__/__/api/c/imageio2.cpp.o
/ufs/eyalroz/src/arrayfire-full-3.3.1/src/api/c/imageio.cpp: In function ‘af_err af_load_image(void**, const char*, bool)’:
/ufs/eyalroz/src/arrayfire-full-3.3.1/src/api/c/imageio.cpp:152:57: error: ‘JPEG_GREYSCALE’ was not declared in this scope
         if(fif == FIF_JPEG && !isColor) flags = flags | JPEG_GREYSCALE;
```
I'm not sure which library this regards, but I do have

```
FREEIMAGE_DYNAMIC_LIBRARY:FILEPATH=/usr/lib64/libfreeimage.so
FREEIMAGE_INCLUDE_PATH:PATH=/usr/include
FREEIMAGE_STATIC_LIBRARY:FILEPATH=FREEIMAGE_STATIC_LIBRARY-NOTFOUND
```
(copied those from my `CMakeCache.txt`), so I should be OK; if not, I should have been notified of something missing.

**Comment by shehzan10** [comment]
This is a freeimage enum that should be there in the header file (/usr/include/FreeImage.h).
Can you also confirm the version of freeimage you are using? You can do this either from the package manager or the version information from the freeimage header file.

**Comment by eyalroz** [comment]
A snippet from `/usr/include/FreeImage.h`:
```
#define FREEIMAGE_MAJOR_VERSION   3
#define FREEIMAGE_MINOR_VERSION   10
#define FREEIMAGE_RELEASE_SERIAL  0
```
And here are all the definitions beginning with JPEG_ in that file:
```
#define JPEG_DEFAULT        0           // loading (see JPEG_FAST); saving (see JPEG_QUALITYGOOD)
#define JPEG_FAST           0x0001      // load the file as fast as possible, sacrificing some quality
#define JPEG_ACCURATE       0x0002      // load the file with the best quality, sacrificing some speed
#define JPEG_CMYK                       0x0004  // load separated CMYK "as is" (use | to combine with other load flags)
#define JPEG_QUALITYSUPERB  0x80        // save with superb quality (100:1)
#define JPEG_QUALITYGOOD    0x0100      // save with good quality (75:1)
#define JPEG_QUALITYNORMAL  0x0200      // save with normal quality (50:1)
#define JPEG_QUALITYAVERAGE 0x0400      // save with average quality (25:1)
#define JPEG_QUALITYBAD     0x0800      // save with bad quality (10:1)
#define JPEG_PROGRESSIVE        0x2000  // save as a progressive-JPEG (use | to combine with other save flags)
```
No enum though.

**Comment by eyalroz** [comment]
Oh, [here's a clue](http://freeimage.sourceforge.net/changeslog.html): It looks like JPEG_GRAYSCALE was added to FreeImage with v3.15.4 , and that Fedora 22, for some odd reason, sticks to 3.10.0.

Well, I guess you need to have some version check in your `CMakeModules/FindFreeImage.cmake`.

**Comment by pavanky** [comment]
Wow, never thought Fedora will have an older package from an older Ubuntu release. We were trying to target versions from Ubuntu 14.04 under the assumption that most other distros will have the packages by now. Looks like Fedora missed out on this for some reason..

**Comment by pavanky** [comment]
@shehzan10 just check for the missing macro and raise an error saying free image on system does not support this image mode.

**Comment by shehzan10** [comment]
The better option is to check `#ifdef JPEG_GREYSCALE`

**Comment by pavanky** [comment]
@shehzan10 that's what I meant.

**Comment by eyalroz** [comment]
With FreeImage 3.17, downloaded from the FreeImage website, the Fedora 22 build concludes, woo-hoo.

Also, instead of just #ifdef JPEG_GRAYSCALE, you could consider doing:

```
  if(fif == FIF_UNKNOWN) {
      AF_ERROR("FreeImage Error: Unknown File or Filetype", AF_ERR_NOT_SUPPORTED);
  }
  int flags = 0;
  if(fif == FIF_JPEG) flags = flags | JPEG_ACCURATE;
- if(fif == FIF_JPEG && !isColor) flags = flags | JPEG_GREYSCALE;
+ if(fif == FIF_JPEG && !isColor) {
+ #ifdef JPEG_GRAYSCALE
+     flags = flags | JPEG_GREYSCALE;
+ #else
+ #warning "the FreeImage library does not define JPEG_GRAYSCALE; non-color JPEG images will not be properly supported.")
+ AF_ERROR("FreeImage Error: Unknown File or Filetype", AF_ERR_NOT_SUPPORTED);
+ #endif
+ }
```

**Comment by pavanky** [comment]
@eyalroz send in a PR!

**Comment by pavanky** [comment]
@eyalroz to `hotfix-3.3.2` not devel.

**Comment by eyalroz** [comment]
@pavanky : I'm not much of a git/github person;  can you spell out what that means? Also, that was a "hand-created diff", I didn't actually check anything out (yet).

**Comment by pavanky** [comment]
@eyalroz nevermind, @shehzan10 tells me he already fixed it his fork and it will be merged into devel soon.

In case you want to contribute in the future, you can follow this document: https://github.com/arrayfire/arrayfire/wiki/Contribute-code-using-github

**Comment by shehzan10** [comment]
Fixed in hotfix-3.3.2.
